File: apps/web/src/types/metric.ts
// Mirrors apps/api/app/schemas/metric.py.

export type MetricType = "quantitative" | "qualitative";
export type MetricStatus = "green" | "amber" | "red" | "grey";

export interface MetricLogEntry {
  at: string;
  by?: string | null;
  by_name?: string | null;
  value: string | null;
  source?: string | null;
  note?: string | null;
  [k: string]: unknown;
}

export interface SuccessMetric {
  id: string;
  account_id: string;
  name: string;
  description: string | null;
  metric_type: MetricType;
  unit: string | null;
  target_value: string | null;
  current_value: string | null;
  status_override: MetricStatus | null;
  log_entries: MetricLogEntry[];
  source: string;
  last_updated_at: string | null;
  last_updated_by: string | null;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  deleted_reason: string | null;
  status: MetricStatus;
  is_editable: boolean;
}

export interface MetricListResponse {
  items: SuccessMetric[];
  total: number;
  is_editable: boolean;
}

export interface MetricCreate {
  name: string;
  description?: string | null;
  metric_type: MetricType;
  unit?: string | null;
  target_value?: string | null;
}

export interface MetricValueLog {
  value: string;
  source?: string | null;
  note?: string | null;
}

export const STATUS_LABELS: Record<MetricStatus, string> = {
  green: "On track",
  amber: "At risk",
  red: "Off track",
  grey: "No data",
};

/** Brand-palette colours per metric status (Beroe brand book Sept 2025).
 *  Values are raw hex so callers can use them in inline styles instead
 *  of the previous Tailwind utility classes — those drifted off-palette
 *  (e.g. green-500 ≠ Risk Green). Mapping:
 *    green → Risk Green  #6EC457
 *    amber → Risk Amber  #F0BC41
 *    red   → Risk Red    #CF4548
 *    grey  → brand mid-grey (Midnight-tinted)
 */
export const STATUS_COLORS: Record<
  MetricStatus,
  { dot: string; bg: string; border: string; text: string }
> = {
  green: { dot: "#6EC457", bg: "#6EC45715", border: "#6EC45740", text: "#1d6b35" },
  amber: { dot: "#F0BC41", bg: "#F0BC4115", border: "#F0BC4140", text: "#854F0B" },
  red:   { dot: "#CF4548", bg: "#CF454810", border: "#CF454830", text: "#7F1D1D" },
  grey:  { dot: "#94a3b8", bg: "#94a3b815", border: "#94a3b830", text: "#475569" },
};

File: apps/api/app/models/metric.py
"""M20 — Success Metric ORM."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, text
from sqlalchemy.dialects.postgresql import ENUM, JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base

MetricType = ENUM(
    "quantitative", "qualitative", name="metric_type", create_type=False
)
MetricStatusOverride = ENUM(
    "green", "amber", "red", "grey", name="metric_status_override", create_type=False
)


class SuccessMetric(Base):
    __tablename__ = "success_metrics"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    account_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("accounts.id", ondelete="CASCADE"), nullable=False
    )

    name: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str | None] = mapped_column(String, nullable=True)
    metric_type: Mapped[str] = mapped_column(MetricType, nullable=False, server_default=text("'quantitative'"))
    unit: Mapped[str | None] = mapped_column(String, nullable=True)
    target_value: Mapped[str | None] = mapped_column(String, nullable=True)
    current_value: Mapped[str | None] = mapped_column(String, nullable=True)
    status_override: Mapped[str | None] = mapped_column(MetricStatusOverride, nullable=True)

    log_entries: Mapped[list] = mapped_column(
        JSONB, nullable=False, server_default=text("'[]'")
    )

    source: Mapped[str] = mapped_column(String, nullable=False, server_default=text("'manual'"))

    last_updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    last_updated_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=text("now()"), nullable=False
    )
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=text("now()"), nullable=False
    )

    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    deleted_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    deleted_reason: Mapped[str | None] = mapped_column(String, nullable=True)
